### PR TITLE
v2

### PR DESCRIPTION
**Pull request by Gummibeer**
The whole update was inspired by:
* https://github.com/hostinger/php-dig
* https://github.com/igorw/dnsresolv

The most notable changes are that the domain is now passed to the `getRecords()` method instead of the construct and to sanitize a domain you can use the dedicated `Domain` class (which is also used internally.
It also uses the native `DNS_*` constants for types but also allows the old names.
The only DNS record type that isn't supported right now is `NAPTR` - I can add it as soon as I have an example domain with one.

### CODE REVIEW

**Comment by Gummibeer** [comment]
@freekmurze I would wait for official PHP8 release and in best case also until StyleCI or PHP-CS-Fixer supports PHP8.

**Comment by freekmurze** [comment]
Ping me when ready! 🙂

**Comment by Gummibeer** [comment]
@freekmurze I've migrated back to PHP7.4 as I haven't used too many PHP8 features that it would be worth waiting for.
Dropping PHP7.4 is possible in a later minor release - so no problem to start v2 with the current PHP support matrix.

**Comment by freekmurze** [comment]
I'll review this soon!

**Comment by Gummibeer** [comment]
Have added all missing tests now. 🎉🚀

**Comment by Gummibeer** [comment]
@freekmurze could you review this PR?

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest]
-        php: ['7.4', '8.0']
+        php: ['8.0', '7.4']
```
Comment (freekmurze) [inline .github/workflows/tests.yml:11]: Let's drop support for PHP 7

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -21,8 +21,7 @@ jobs:
       uses: shivammathur/setup-php@v2
       with:
         php-version: ${{ matrix.php }}
-        tools: composer
-        coverage: pcov
+        tools: composer:v2
```
Comment (freekmurze) [inline .github/workflows/tests.yml:24]: Wouldn't this be the default?

File: composer.json [hunk 3]
```diff
@@ -19,15 +19,25 @@
...
+        {
+            "name": "Tom Witkowski",
+            "email": "dev.gummibeer@gmail.com",
+            "homepage": "https://gummibeer.de",
+            "role": "Developer"
         }
     ],
     "require": {
-        "php": "^7.4|^8.0",
+        "php": "^7.4 || ^8.0",
```
Comment (freekmurze) [inline composer.json:31]: Let's only support PHP 8

File: src/Dns.php [hunk 4]
```diff
@@ -2,139 +2,119 @@
...
+    protected ?Factory $factory;
 
-        $this->nameserver = $nameserver;
+    protected ?string $nameserver = null;
+
+    public function __construct(
+        ?Types $types = null,
+        ?Factory $factory = null
+    ) {
+        $this->types = $types ?? new Types();
```
Comment (freekmurze) [inline src/Dns.php:28]: We could use the null coalescing assignment operator here: `??=`

File: src/Dns.php [hunk 5]
```diff
@@ -2,139 +2,119 @@
...
         return $this->nameserver;
     }
 
-    public function getRecords(...$types): string
+    /**
+     * @param \Spatie\Dns\Support\Domain|string $search
+     * @param int|string|string[] $types
+     * @return \Spatie\Dns\Contracts\Collection
+     */
+    public function getRecords($search, $types = DNS_ALL): CollectionContract
```
Comment (freekmurze) [inline src/Dns.php:50]: If we use union types, we can lose the docblock

File: src/Handlers/Handler.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+use Spatie\Dns\Records\Record;
+use Spatie\Dns\Support\Factory;
+
+abstract class Handler
+{
+    protected ?string $nameserver = null;
+
+    protected Factory $factory;
+
+    public function __construct(Factory $factory)
```
Comment (freekmurze) [inline src/Handlers/Handler.php:14]: Let's use PHP 8 promoted properties wherever possible 👍

File: src/Records/CAA.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            'type' => $attributes[3],
+            'flags' => $attributes[4],
+            'tag' => $attributes[5],
+            'value' => $attributes[6],
+        ]);
+    }
+
+    public function __toString(): string
+    {
+        return sprintf(
```
Comment (freekmurze) [inline src/Records/CAA.php:33]: I think using regular string concatenation would be more readable.

File: src/Records/MX.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            $this->host,
+            $this->ttl,
+            $this->class,
+            $this->type,
+            $this->pri,
+            $this->target
+        );
+    }
+
+    protected function castPri($value): string
```
Comment (freekmurze) [inline src/Records/MX.php:41]: It's not clear to me what `pri` means? Primitive? Let's not use abbreviations in method names.

File: src/Records/CNAME.php [hunk 9]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Spatie\Dns\Records;
+
+/**
+ * @method string target()
+ */
+class CNAME extends Record
```
Comment (freekmurze) [inline src/Records/CNAME.php:8]: Even though the types seem to be uppercased in DNS, I would like to follow regular class capitalisation conventions. So let's rename this to `Cname`. Rename the others in a similar fashion.

File: src/Records/MX.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace Spatie\Dns\Records;
+
+/**
```
Comment (freekmurze) [inline src/Records/MX.php:5]: Why do we have these docblocks on this class and similar classes? The methods seem to exists on the classes.

**Review by freekmurze — CHANGES_REQUESTED** [review]
I've done a shallow review now, I'll do another one when you've addressed the feedback in this review.

In case users should use this package differently then before, could you also update the readme?

File: src/Records/MX.php (hunk 10 above)
Comment (Gummibeer) [reply src/Records/MX.php:5]: No, the property exists but the method is called by `__call()` on the `Record` base class.

File: src/Records/MX.php (hunk 8 above)
Comment (Gummibeer) [reply src/Records/MX.php:41]: I've used the exact names like in DNS and plain PHP (also for mapping) in this case `pri` is the MX record priority.
If I change this I will have to add a naming mapping and can't use the current auto/magic logic.

File: src/Records/CAA.php (hunk 7 above)
Comment (Gummibeer) [reply src/Records/CAA.php:33]: I think regular concatenation not but string interpolation could be helpful.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (Gummibeer) [reply .github/workflows/tests.yml:24]: yes, in time I've created this it wasn't.^^

File: src/Records/MX.php (hunk 8 above)
Comment (freekmurze) [reply src/Records/MX.php:41]: Ok, let's keep it like it is

File: .github/workflows/tests.yml (hunk 2 above)
Comment (freekmurze) [reply .github/workflows/tests.yml:24]: The world is a better place now 🙂

File: src/Records/CNAME.php (hunk 9 above)
Comment (Gummibeer) [reply src/Records/CNAME.php:8]: I totally get you. Just think that `Mx` or `Aaaa` look even worse. 🤔
I will do it if you want. Just want to say that most won't really benefit from regular class naming.
And the `A` class name also doesn't really follow normal conventions.

**Comment by Gummibeer** [comment]
I've addressed all requests except the class naming because of my last comment.

**Comment by freekmurze** [comment]
Thanks! I'll review and polish it on our local `v2` branch.

**Pull request merged**